internals/src/error/input_string.rs

```diff
@@ -0,0 +1,126 @@
+//! Implements the [`InputString`] type storing the parsed input.
+
+use core::fmt;
+
+use storage::Storage;
+
+/// Conditionally stores the input string in parse errors.
+///
+/// This type stores the input string of a parse function depending on whether `alloc` feature is
+/// enabled. When it is enabled, the string is stored inside as `String`. When disabled this is a
+/// zero-sized type and attempt to store a string does nothing.
+///
+/// This provides two methods to format the error strings depending on the context.
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct InputString(Storage);
+
+impl InputString {
+    /// Displays a message saying `failed to parse <self> as <what>`.
+    ///
+    /// This is normally used whith the `write_err!` macro.
+    pub fn display_cannot_parse<'a, What>(&'a self, what: &'a What) -> CannotParse<'a, What>
```

Comment:
Commenting because this is the second time I reviewed this code and both times I did a double take at the `What` generic, I personally prefer a plain old `T` so I don't go "what is `What`" then have to look back at `<'a, What`, and go "oh its a generic with a name, that's unexpected". libp2p does this, it used to drive me nuts there too. Open to expanding my world view if you have a valid technical reason to back it though :)

Reply:
Lately I started using bit more descriptive names for generics but it's true it's not clear it's generic. Hard to say what's better. We could use some silly prefix like `GWhat` but I'd rather not.

Another good name for `What` would be `Object` but it could look like OOP object while it actually refers to a word in a phrase (same category as "subject").

Reply:
None of these make the code easier to read than plain old `T` in my opinion.

Reply:
I found it helpful for more complicated code but this is simple, so I don't care.